liquidity_pool_router_with_payload/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+        );
+
+        pool_client.deposit(&to);
+    }
+}
+
+pub trait PayloadTrait {
+    fn has_sig(e: Env, function: Symbol) -> bool;
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)>;
+}
+
+pub struct LiquidityPoolRouterPayload;
+
+#[contractimpl]
+impl PayloadTrait for LiquidityPoolRouterPayload {
+    fn has_sig(_e: Env, function: Symbol) -> bool {
+        const DEPOSIT_RAW: u64 = symbol!("deposit").to_raw().get_payload();
+        match function.to_raw().get_payload() {
+            DEPOSIT_RAW => true,
+            _ => false,
+        }
+    }
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)> {
+        const DEPOSIT_RAW: u64 = symbol!("deposit").to_raw().get_payload();
+        match function.to_raw().get_payload() {
+            DEPOSIT_RAW => {
+                let to: Identifier = args.get_unchecked(0).unwrap().try_into_val(&e).unwrap();
+                let deposit_a: Deposit = args.get_unchecked(1).unwrap().try_into_val(&e).unwrap();
+                let deposit_b: Deposit = args.get_unchecked(2).unwrap().try_into_val(&e).unwrap();
+
+                let pool_id = get_pool_id(&e, &pool_salt(&e, &deposit_a.token, &deposit_b.token));
+                let reserves = pool_contract::Client::new(&e, &pool_id).get_rsrvs();
+                let amounts = get_deposit_amounts(
+                    deposit_a.desired,
+                    deposit_a.min,
+                    deposit_b.desired,
+                    deposit_b.min,
+                    reserves,
+                );
+
+                let mut callstack_a = callstack.clone();
+                callstack_a.push_back((deposit_a.token.clone(), symbol!("xfer")));
+                let to_forward_a = token_contract::Client::new(&e, &deposit_a.token)
+                    .payload(
+                        &symbol!("xfer"),
+                        &(&to, deposit_a.nonce, e.get_current_contract(), amounts.0).into_val(&e),
+                        &callstack_a,
+                    )
+                    .get_unchecked(symbol!("sig"))
+                    .unwrap();
+
+                let mut callstack_b = callstack.clone();
+                callstack_b.push_back((deposit_b.token.clone(), symbol!("xfer")));
+                let to_forward_b = token_contract::Client::new(&e, &deposit_b.token)
+                    .payload(
+                        &symbol!("xfer"),
+                        &(&to, deposit_b.nonce, e.get_current_contract(), amounts.1).into_val(&e),
+                        &callstack_b,
+                    )
+                    .get_unchecked(symbol!("sig"))
+                    .unwrap();
+
+                let to_verify = SaltedSignaturePayload::V0(SaltedSignaturePayloadV0 {
```

Comment:
I think this isn't quite right. I'm pretty sure I would need to add `to_forward_a` and `to_forward_b` to this signature payload.